File: ingestion/social.py
import os
import json
import httpx
from datetime import datetime
from typing import List, Literal
from ingestion.models import Atom, SocialPost, VerificationResult
from ingestion.verification import verify_post_against_atoms

def generate_social_post(
    atoms: List[Atom], 
    platform: Literal["Twitter", "LinkedIn"],
    tone: str = "professional"
) -> SocialPost:
    """
    Generates a social media post from a list of atoms and verifies it.
    """
    groq_key = os.getenv("GROQ_API_KEY")
    groq_model = os.getenv("GROQ_MODEL") or "mistral-7b-instruct"
    
    if not groq_key:
        raise ValueError("GROQ_API_KEY not found in environment.")

    # Prepare context
    context = "\n---\n".join([f"Atom {a.atom_id}: {a.text}" for a in atoms])
    
    system = (
        f"You are a Social Media Manager. Create a {tone} {platform} post based on the provided news atoms. "
        "The post should be engaging and accurate. "
        "Return ONLY the post content as a raw string."
    )
    user = (
        f"News Atoms:\n{context}\n\n"
        f"Create a {platform} post."
    )

    payload = {
        "model": groq_model,
        "messages": [
            {"role": "system", "content": system},
            {"role": "user", "content": user},
        ],
        "temperature": 0.7,
    }
    headers = {
        "Authorization": f"Bearer {groq_key}",
        "Content-Type": "application/json",
    }

    with httpx.Client(timeout=30.0) as client:
        r = client.post(
            "https://api.groq.com/openai/v1/chat/completions",
            headers=headers,
            json=payload,
        )
        r.raise_for_status()
        data = r.json()
    
    post_content = data["choices"][0]["message"]["content"].strip()
    
    # Run verification immediately
    verification = verify_post_against_atoms(post_content, atoms)
    
    # Generate a stable ID (mocking stable_id for now as I don't want to import utils yet if not needed)
    import hashlib
    post_id = hashlib.md5(post_content.encode()).hexdigest()[:12]
    
    return SocialPost(
        post_id=post_id,
        platform=platform,
        content=post_content,
        source_atom_ids=[a.atom_id for a in atoms],
        verification=verification,
        created_at=datetime.now()
    )

File: lib/social.ts
import { generateJSON, generateText } from "./llm";
import { type Atom, type SocialPost, type VerificationResult } from "./types";
import crypto from "node:crypto";

export async function generateSocialPost(
  atoms: Atom[],
  platform: "Twitter" | "LinkedIn",
  tone: string = "professional"
): Promise<SocialPost> {
  const context = atoms.map((a) => `[Source ${a.atom_id}]: ${a.text}`).join("\n---\n");

  const prompt = `You are a Social Media Manager. Create a ${tone} ${platform} post based on the following news atoms. 
  The post should be engaging and accurate.
  
  News Atoms:
  ${context}
  
  Return ONLY the post content.`;

  const content = await generateText({
    prompt,
    temperature: 0.7,
    maxTokens: 1024,
  });
  const post_id = crypto.createHash("md5").update(content).digest("hex").slice(0, 12);

  // Auto-verify
  const verification = await verifyPost(content, atoms);

  return {
    post_id,
    platform,
    content,
    source_atom_ids: atoms.map((a) => a.atom_id),
    verification,
    created_at: new Date().toISOString(),
  };
}

export async function verifyPost(content: string, atoms: Atom[]): Promise<VerificationResult> {
  const context = atoms.map((a) => `[Source ${a.atom_id}]: ${a.text}`).join("\n---\n");

  const system = `You are a Fact-Checking Agent. Verify the discrete claims in a Social Media Post against the provided Source Atoms.
  Return ONLY JSON with the following structure:
  {
    "is_true": boolean,
    "score": number (0 to 1),
    "reasoning": string,
    "supported_claims": string[],
    "unsupported_claims": string[],
    "conflicting_atom_ids": string[]
  }`;

  const user = `Social Media Post:
  ${content}
  
  Source Atoms:
  ${context}
  
  Verify claims and return JSON only.`;

  try {
    const result = await generateJSON<any>({
      system,
      prompt: user,
      maxTokens: 2048,
    });
    return {
      is_true: result.is_true ?? (result.score > 0.7),
      score: result.score ?? 0,
      reasoning: result.reasoning ?? "",
      supported_claims: result.supported_claims ?? [],
      unsupported_claims: result.unsupported_claims ?? [],
      conflicting_atom_ids: result.conflicting_atom_ids ?? [],
    };
  } catch (e) {
    return {
      is_true: false,
      score: 0,
      reasoning: "Verification failed to parse JSON response.",
      supported_claims: [],
      unsupported_claims: ["System Error"],
      conflicting_atom_ids: [],
    };
  }
}
